tools/shared.py

```diff
@@ -772,6 +772,54 @@ def clean_temp():
     prepare_to_clean_temp(EMSCRIPTEN_TEMP_DIR) # this global var might change later
   return EMSCRIPTEN_TEMP_DIR
 
+class Enum(set):
+  def __getattr__(self, key):
+    if key in self:
+      return key
+
+WarningType = Enum(['ABSOLUTE_PATHS', 'SEPARATE_ASM', 'ALMOST_ASM'])
```

Comment:
Come to think of it, it seems a little funny that we are enumerating all of the warning types (and constructing their corresponding flag names) here, but we are are defining the text elsewhere. It seems like it should be one of the following cases:
1. there is no central list of all the warnings (in which case both the warning/flag name and warning text are passed to `warn()`). Then `WarningManager` just becomes a utility class that formats the message and tracks whether warnings are enabled and printed, and handles flags based on names/enums. The external code defines which warnings exist, their names, and the conditions under which they fire.
2. WarningManager has the list of all warnings, along with their enums/names and warning text. Then the external code only determines the conditions under which they fire.

2 would seem to be better if there is more than one condition or external location that can cause a warning (it would avoid duplication), but that doesn't seem to be the case currently. Do either of you have any intuition about whether that will be the case in the future?